nuxeo-features/nuxeo-platform-comment/nuxeo-platform-comment/src/test/java/org/nuxeo/ecm/platform/comment/AbstractTestAnnotationNotification.java

```diff
@@ -153,23 +154,67 @@ protected void captureAndVerifyAnnotationEventNotification(Supplier<DocumentMode
             assertEquals(1, handledEvents.size());
             Event expectedEvent = handledEvents.get(0);
             assertEquals(annotationEventType, expectedEvent.getName());
-            checkDocumentEventContext(expectedEvent, annotationDocumentModel, annotatedDocumentModel);
+
+            checkDocumentEventContext(expectedEvent, annotationDocumentModel, annotatedDocumentModel,
+                    annotatedDocumentModel);
             checkReceivedMail(emailsResult.getMails(), annotationDocumentModel, annotatedDocumentModel,
                     handledEvents.get(0), annotationEventType);
         }
     }
 
+    @Test
+    public void shouldNotifyWithTheRightAnnotatedDocument() {
+        try (CapturingEventListener listener = new CapturingEventListener(COMMENT_ADDED, DOCUMENT_CREATED)) {
+            Comment createdAnnotation = createAnnotationAndAddSubscription("CommentAdded", "Creation");
+            DocumentModel annotationDocumentModel = session.getDocument(new IdRef(createdAnnotation.getId()));
+            transactionalFeature.nextTransaction();
+
+            assertTrue(listener.hasBeenFired(COMMENT_ADDED));
+            assertTrue(listener.hasBeenFired(DOCUMENT_CREATED));
+
+            List<Event> handledEvents = listener.streamCapturedEvents()
+                                                .filter(e -> COMMENT_ADDED.equals(e.getName()))
+                                                .collect(Collectors.toList());
+
+            assertEquals(1, handledEvents.size());
+
+            checkDocumentEventContext(handledEvents.get(0), annotationDocumentModel, annotatedDocumentModel,
+                    annotatedDocumentModel);
+            checkReceivedMail(emailsResult.getMails(), annotationDocumentModel, annotatedDocumentModel,
+                    handledEvents.get(0), COMMENT_ADDED);
+
+
+            Comment reply = createAnnotation(annotationDocumentModel);
+            DocumentModel replyDocumentModel = session.getDocument(new IdRef(reply.getId()));
+            transactionalFeature.nextTransaction();
+
+            handledEvents = listener.streamCapturedEvents()
+                    .filter(e -> COMMENT_ADDED.equals(e.getName()))
+                    .collect(Collectors.toList());
+            assertEquals(2, handledEvents.size());
+
+            checkDocumentEventContext(handledEvents.get(1), replyDocumentModel, annotationDocumentModel,
+                    annotatedDocumentModel);
+            checkReceivedMail(emailsResult.getMails(), replyDocumentModel, annotatedDocumentModel,
+                    handledEvents.get(1), COMMENT_ADDED);
+        }
+    }
```

Comment:
i took  some time to see how we reuse the exiting code to make the checks as here you do almost the same things / checks that are on the `captureAndVerifyAnnotationEventNotification` and you changed the `org.nuxeo.ecm.platform.comment.CommentUtils#checkReceivedMail` with a new control
```
 DocumentModel sourceDocument = ((DocumentEventContext) event.getContext()).getSourceDocument();

        if (COMMENT_ADDED.equals(commentEventType) && !commentedDocModel.getId().equals(sourceDocument.getId())) {
            expectedMailsCount = 2;
        }
```
i am wondering if we can rework this test which will allow us to add more coverture on `update`, `remove`...
and revert the modification on `org.nuxeo.ecm.platform.comment.CommentUtils#checkReceivedMail`
here how i modified your test 
```
@Test
    public void shouldNotifyWithTheRightAnnotatedDocument() {
        // First comment
        captureAndVerifyAnnotationEventNotification(() -> {
            Annotation createdAnnotation = createAnnotationAndAddSubscription("CommentAdded", "Creation");
            return session.getDocument(new IdRef(createdAnnotation.getId()));
        }, COMMENT_ADDED, DOCUMENT_CREATED);

        // Reply
        captureAndVerifyAnnotationEventNotification(() -> {
            Comment reply = createAnnotation(annotatedDocumentModel);
            DocumentModel replyDocumentModel = session.getDocument(new IdRef(reply.getId()));
            return session.getDocument(new IdRef(replyDocumentModel.getId()));
        }, COMMENT_ADDED, DOCUMENT_CREATED);
    }
```

doing just this is not enough because `emailsResult.getMails()` will returns the whole emails since the start of the tests, what we need is to just keep the real emails sent by the test i made some changes on `captureAndVerifyAnnotationEventNotification` as below 
```
..........
     // Deduct only the emails sent since the last time
            List<SmtpMailServerFeature.MailMessage> previousMails = new ArrayList<>(mails);
            mails.clear();
            mails.addAll(emailsResult.getMails()
                    .stream()
                    .filter(Predicate.not(previousMails::contains))
                    .collect(Collectors.toList()));

            checkReceivedMail(mails, annotationDocumentModel, annotatedDocumentModel,
                    handledEvents.get(0), annotationEventType);
``` 
 the thing is i introduced a new field ` protected List<SmtpMailServerFeature.MailMessage> mails = new ArrayList<>();` that we can clean properly at the end of the whole class test, wdyth, we can ask the team for more feedbacks ...

Reply:
Yes this is the first thing i tried. But when I tried to empty the previous mails I couldn't get it to work.
If this works then fine for me.

Reply:
i did it and it works and i launched the whole test on annotation notification and they are ok -> but we should check the other stack test, you can take the code that i past on the snippet above

Reply:
This has been reworked again when removing all the mail filtering. Now subscription is done after the first comment so we don't need to filter.